### ISSUE TITLE
Removing Drawer dialog in Mobile Mode?

### ISSUE THREAD

**Issue by frastlin** [vaadin/vaadin-app-layout#165]
Hello,
When in a mobile mode, the tabs go into a dialog that is labeled "drawer". Sorry, but "drawer" is probably the most uninformative name one could think of for this dialog. First, if the dialog part could be made a role="presentation", that may help, as it's extremely weird to have tabs in a random dialog. Second, If there is no way to get rid of the dialog, please call it something useful like "Navigation" or "tabs".
Screen reader users often don't realize their windows are not expanded, and they often get put into the mobile view on the desktop. This means that the site may be different in two different tabs, and it's extremely confusing.
Thanks,

**Comment by rolfsmeds** [comment]
Thanks for the ticket, but could you rephrase the problem and solution proposal with a bit more clarity and a bit less rant? We really do want to get a11y issues fixed (and have an ongoing project to identify and analyze them), but tickets written like this make that more difficult than it needs to be.

It sounds to me that you're describing two different problems here, but I'm not quite sure. If they're different aspects of the same problem, it's fine to have them in a single ticket, otherwise it would be better to split them into separate tickets.

So:
- What are the problems for the end user?
- What are your proposed fixes?

**Comment by jouni** [comment]
Thanks for raising the issue!

We use “drawer” because that has become an established name for this UI pattern, pioneered by Material Design, if I remember correctly: https://material.io/components/navigation-drawer

If I understand correctly, the problem is that even though the drawer is visually hidden, it’s still possible for the user to move keyboard focus there using the tab key. That is indeed an issue that we should fix.

An additional fix could be that when the drawer is in the “overlay” mode and shows the backdrop/curtain when it’s open, it is implying it would be modal. We might want to contain keyboard focus within the drawer in that case as well. Also, when the user opens the overlay drawer, we should move keyboard focus to it, and move it back to the drawer toggle when the user dismisses the drawer. If they choose to navigate somewhere else in the app, the focus should probably be moved to the “content area” of the application.

**Comment by frastlin** [comment]
Hello,
Yes, the problem is there are a lot of problems.

1. The first fix is to remove lines 99, 100, and 101 from:
https://github.com/vaadin/vaadin-app-layout/blob/next/src/vaadin-app-layout-mixin.ts
This removes the idea of a drawer, removes the dialog, and makes telling a user a dialog is a "drawer" moot. This fix can be it, screen reader users will just see the drawer as a set of always expanded tabs.
2. To get the behavior of a drawer, we need to have the [Details](https://vaadin.com/components/vaadin-details/html-examples) element as the drawer toggle, which would fix [this issue,](https://github.com/vaadin/vaadin-app-layout/issues/164). Along with the fix in 1, this would add the behavior we need.
3. Before this will work, the problem with the [Vaadin button element](https://github.com/vaadin/vaadin-button/issues/150) needs to be fixed.

All a drawer to a screen reader is a Details element, where you have a button that is collapsed, and when you click it, the button becomes expanded, and content shows.

@jouni is correct, when the user clicks a tab, they should be [jumped to the content.](https://github.com/vaadin/flow/issues/9076#issuecomment-700628350)

**Comment by frastlin** [comment]
Also, most screen reader navigation uses the arrow keys, and tab is only used on forms, and when navigating quickly through a page, so not extremely relevant, other than either all the links in the drawer be put in the tab index, or only the focused tab has an index, and the other links are removed, with the next tab focusing on the content. Most screen reader users would prefer using the key command to jump to the h1 on the page, and when they want to navigate, hit ctrl+home and press the down arrow to get to the tabs.

**Comment by jouni** [comment]
If I may ask, are you a screen reader user yourself, or do you happen to know some? We really appreciate if you have real-world experience or even second-hand knowledge of actual screen reader user behavior. We unfortunately don’t have those, and rely on information from the internet, user feedback, and our own rudimentary knowledge of using screen readers.

**Comment by frastlin** [comment]
Yes, I'm a screen reader user who does Javascript development.

**Comment by rolfsmeds** [comment]
This was extremely useful, thanks.

We need to remove the hidden drawer from tab-index for non-SR users regardless, as otherwise non-SR users' keyboard focus will keep ending up in an element that is not visible to them.

Should it then also be hidden from SR-users, for consistency? My initial thought was that it could always be available for SR-users, as the point of an overlay drawer is mainly to conserve screen real estate, which is meaningless for SR-users, but I'm wondering if this is problematic somehow.

If the drawer's content is always visible, I suppose we could simply hide the toggle button from SR-users?

**Comment by frastlin** [comment]
@rolfsmeds  the developer adds the element for the button that shows and hides the drawer, I just put an aria-hidden on it, and that worked.
It's not completely true that drawers are not useful for screen reader users, if there is a long menu, and the SR user does not want to read through it, and there is not a main landmark or heading 1, then it's nice to close the menu. It's also hard for a SR developer to understand what the drawer does if they are not seeing the expand and collapse functionality.
The option that gives the SR and keyboard users the experience that is intended is to use the Details element, and make the selected tab and the expand/collapse button tabbable (give them a tabindex), and remove the other tabs from the tabindex.
These other options are not a drawer experience, so should not be in the drawer option. It's always better to give SR and non SR users the same experience when possible, and adding a details element is completely possible. The default name for the expand/collapse button should be "Menu" or "Tabs", as either of those are in the common vernacular for users.

**Comment by frastlin** [comment]
BTW, the Vaadin components are something I would like to see have an amazing experience, as it's the only library, that's not accessibility focused, with a working grid, tab, details, accordion, dialog, and menu element.
It's just these few elements, along with the very basic elements, that need some fixing.
